Jip 2 api 연결 프론트

BHyeonKim opened the pull request:
# PR

## 🗄️ 종류

PR 종류를 선택하세요

- [ ] Code Review
- [x] New Feature
- [ ] Remove Feature
- [ ] Change Logic
- [ ] Bug Fix
- [x] Setup

<br>

## 🔍 왜 변경을 했습니까?

- 로그인 기능을 구현하였습니다.
- 토큰등의 인증 정보를 리덕스와 asyncStorage에 저장하였습니다.

<br>

## 🔧 작업 내용

- @react-native-async-storage/async-storage, dotenv, expo-constants 설치
- 환경변수를 숨기기 위한 패키지 설치 -  .env을 사용하기 위해 `dotenv, expo-constants`을 설치하였습니다.
추가적으로 .env 파일의 내용을 가져오고 사용하기 위해

```
app.config.ts
config.ts
```
를 작성하고
`custom.d.ts`에
```
declare module 'react-native-dotenv' {
  export const OAUTH_KAKAO_URI: string
}
```
를 추가했습니다.
<br/>
- 인증정보를 저장하기 위한 슬라이스 작성
- 로그인 로직 작성
사용자가 버튼을 누르면 로그인을 위한 웹뷰 화면으로 라우팅됩니다.
후에 서버에서 주는 토큰과 회원정보입력여부를 리덕스에 저장합니다.
App.tsx에서는 스토어의 변화를 보고 async-storage에 인증정보를 업데이트 합니다.
```
store.subscribe(() => {
  const { auth } = store.getState()
  ;(async () => {
    await setAuthInfo(auth)
  })()
})
```
앱의 실행시에는 App.tsx의 마운트 후에 async-storage에 있는 값을 리덕스에 전달합니다.
```
  useEffect(() => {
    getAuthInfo().then((authInfo) => {
      if (!authInfo) return
      dispatch(login(authInfo))
    })
  }, [dispatch])
```
<br/>
- 쿼리스트링 파싱함수 작성 

```
export const parseQueryString = (
  url: string
): null | { [key: string]: string } => {
```

인증정보 파싱(token:string, hasInfo:boolean) 뿐만이 아닌 모든 쿼리스트링에 유니버셜하게 사용하고 싶어서 리턴 타입은 `{ [key: string]: string }`로 두었습니다.

<br>

## 📷 스크린샷

<br>

## 📝 리뷰할 때 참고할 점

- App.tsx에서 IIFE를 사용하기 위해 `/* eslint-disable @typescript-eslint/no-floating-promises */` 를 추가했습니다.
- protected routes는 [공식문서](https://reactnavigation.org/docs/auth-flow/)를 참고했습니다.

BHyeonKim committed:
feat(native): 로그인용 웹뷰 라우트 추가

JIP-2

BHyeonKim committed:
feat(native): 쿼리파라미터 파싱용 함수

JIP-2

BHyeonKim committed:
feat(native): .env 세팅

+expo-constants, app.config.ts, config.ts 파일 작성

JIP-2

BHyeonKim committed:
feat(native): 로그인용 WebView 스크린 작성

JIP-2

BHyeonKim committed:
feat(native): 리덕스 로그인, 로그아웃 로직 구현

JIP-2

BHyeonKim committed:
chore(native): react-native-async-storage/async-storage 설치

JIP-2

BHyeonKim committed:
refactor(native): auth 타입 분리

JIP-2

BHyeonKim committed:
refactor(native): nesting된 삼항연산자 분리

JIP-2

BHyeonKim committed:
feat(native): async 스토리지에 토큰 및 정보입력 여부 저장

async storage에 토큰 및 정보입력 여부 객체등의 인증정보를 저장했습니다. redux의 store을 구독하여 store에 변화가 생길 시 리덕스에 저장된 인증정보를
asyncstorage에 전달합니다. 앱이 켜저 Native 컴포넌트가 마운트가 되면 asyncstorage에 있는 인증 정보를 다시 리덕스에 전달합니다.

JIP-2

BHyeonKim committed:
chore(native): console.log 삭제

jkpark104 commented on apps/native/src/utils/parseQuery.ts:
```diff
@@ -0,0 +1,17 @@
+export const parseQueryString = (
+  url: string
+): null | { [key: string]: string } => {
+  const [_, queryString] = url.split('?')
```
const [, queryString] = url.split('?')  이렇게 표현하면 eslint disabled하지 않아도 될 것 같습니다

const queryString = url.split('?')[1] <-- 요것도 괜찮구요 👍

BHyeonKim replied:
아, 아예 안받아도 되군요...!
감사합니다!

jkpark104 commented on apps/native/src/utils/asyncStorage.ts:
```diff
@@ -0,0 +1,21 @@
+    const jsonValue = JSON.stringify(value)
+    await AsyncStorage.setItem('@storage_Key', jsonValue)
+  } catch (e) {
+    // saving error
```
catch 문 안에 console.error나 throw error로 표시하는 건 어떠신가요?

BHyeonKim replied:
넵 알겠습니다!.

jkpark104 commented on apps/native/src/utils/parseQuery.ts:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable @typescript-eslint/naming-convention */
+export const parseQueryString = (
```
질문이 있습니다. 🤔 요 함수가 필요한 이유는 네이티브에서는 --> new URLSearchParams 함수를 사용할 수 없어서 일까요?

BHyeonKim replied:
넵, 리엑트 네이티브에서는 사용하려면 라이브러리를 설치해야 하더라구요.
그래서 그냥 구현했습니다😭
[참조](https://velog.io/@ehgks0000/%E3%85%81%E3%84%B4%E3%85%87%E3%84%B9-kb09e0rf)

jkpark104 commented on apps/native/src/App.tsx:
```diff
@@ -1,12 +1,33 @@
+  const { auth } = store.getState()
+  ;(async () => {
+    await setAuthInfo(auth)
+  })()
```
즉시실행함수를 사용하지 않고 `setAuthInfo(auth)`와 동작이 동일하다고 생각하는데 즉시실행과 await를 사용하신 이유가 있을까요?

BHyeonKim replied:
헉 지금 보니깐 사용한 이유가 없었군요...
`setAuthInfo(auth)`로 고쳤습니다.
뒤에  `getAuthInfo().then`에서 await을 안 사용한다고 린터가 지적하여 규칙제외 주석은 남겨두었습니다. 😄

jkpark104 replied:
저도 비슷한 문제가 있었는데 catch도 같이 있으면 에러 없어지는 것 같았습니다!

BHyeonKim committed:
refactor(native): 코드리뷰 사항 반영

JIP-2

BHyeonKim committed:
refactor(native): 토큰을 저장하고 가저오는 로직 수정

store.subscribe를 사용하면 엑션이 발생할 때마다 async-storage를 업데이트하는 setAuthInfo함수를 호출하여 퍼포먼스 이슈가 발생할 것 같습니다. 따라서
useEffect를 사용하여 마운트 직후에는 실행시키지 않고 auth가 변경되면 async-storage를 업데이트 하도록 로직을 바꾸었습니다.

JIP-2

BHyeonKim commented:
store.subscribe를 사용하면 엑션이 발생할 때마다 async-storage를 업데이트하는 setAuthInfo함수를 호출하여 퍼포먼스 이슈가 발생할 것 같습니다.
제가 회원정보입력에서 리덕스를 사용하기 때문에 쓸데없는 함수 실행이 일어납니다.
따라서 subscribe를 사용하지 않는 쪽으로 로직을 변경하였고, `useUpdateAuth`라는 이름의 커스텀 훅으로 분리했습니다.

jkpark104 reviewed (approved):
고생하셨습니다 👍
